File: fusion/transforms.py
"""Per-model prompt transformation — deliberation → family-native format.

Maps each panelist model to its family (claude/gpt/glm/qwen), renders the
canonical task_spec into a system+user prompt using the family's template,
and selects per-family sampling params.

Research sources:
  claude: docs.claude.com/prompt-engineering (XML tags, effort, no prefilling on 4.6+)
  gpt:    cookbook.openai.com/gpt-5 (reasoning model — do NOT prompt CoT, reasoning_effort)
  glm:    github.com/zai-org/glm-5 (OpenAI-compatible, repetition_penalty 1.05, no json_schema)
  qwen:   qwen.readthedocs.io (ChatML, non-thinking temp 0.7/top_p 0.8, never greedy)
"""

from pathlib import Path
from typing import Any

PROMPTS_DIR = Path(__file__).resolve().parent / "prompts"

# Per-family sampling defaults (sourced from official docs).
# Variant-specific overrides (opus→xhigh, haiku→medium, mini→medium) applied in transform.
FAMILY_PARAMS: dict[str, dict[str, Any]] = {
    "claude": {"effort": "high"},        # default; opus→xhigh, haiku→medium
    "gpt":    {"reasoning_effort": "medium"},  # default; gpt-5.5→high, mini→medium
    "glm":    {"temperature": 0.7, "top_p": 0.9, "repetition_penalty": 1.05},
    "qwen":   {"temperature": 0.7, "top_p": 0.8},  # non-thinking mode for 27B
}

# Model variant → effort/reasoning_effort overrides (within a family)
VARIANT_EFFORT: dict[str, str] = {
    "claude-opus-4-8":   "xhigh",
    "claude-sonnet-4-6": "high",
    "claude-haiku-4-5":  "medium",
    "gpt-5.5":           "high",
    "gpt-5.4":           "medium",
    "gpt-5.4-mini":      "medium",
    "gpt-5.3-codex":     "high",
    "gpt-5.2":           "medium",
}


def family_of(model: str) -> str:
    """Detect model family from the model slug.

    Order matters: qwopus contains "opus" so qwen must be checked before claude.
    """
    m = model.lower()
    if "qwen" in m or "qwopus" in m or "ollama" in m or "turboquant" in m:
        return "qwen"
    if "claude" in m or "opus" in m or "sonnet" in m or "haiku" in m:
        return "claude"
    if "gpt" in m:
        return "gpt"
    if "glm" in m:
        return "glm"
    # ponytail: unknown → claude-style (most general XML format) — upgrade: add explicit fallback policy
    return "claude"


def _coerce_str(value: Any) -> str:
    """Render task_spec values to template strings (lists → bullet lines, None → empty)."""
    if value is None:
        return ""
    if isinstance(value, list):
        return "\n".join(f"- {item}" for item in value)
    return str(value)


def load_family_template(family: str) -> str:
    """Load the .tmpl file for a family. Raises FileNotFoundError if missing."""
    path = PROMPTS_DIR / "formats" / f"{family}.tmpl"
    return path.read_text()


def _split_system_user(rendered: str) -> tuple[str, str]:
    """Split a rendered template into (system, user) by the [SYSTEM]/[USER] markers."""
    # ponytail: simple marker split — upgrade: proper template parser if complexity grows
    if "[USER]" in rendered:
        parts = rendered.split("[USER]", 1)
        system_block = parts[0]
        user = parts[1].strip()
    else:
        system_block = rendered
        user = ""

    # Strip the [SYSTEM] marker from the system block
    system = system_block.replace("[SYSTEM]", "", 1).strip()
    return system, user


def transform_for_model(
    task_spec: dict[str, Any],
    model: str,
) -> tuple[str, str, dict[str, Any]]:
    """Render the canonical task_spec into (system, user, params) for the model's family.

    task_spec keys: role, task, context, constraints, output_format, guidance_blocks, input
    Returns (system_prompt, user_prompt, sampling_params).
    """
    family = family_of(model)
    template = load_family_template(family)

    # Build the substitution map — coerce all values to strings
    subs = {
        "role":            _coerce_str(task_spec.get("role", "an expert assistant")),
        "role_expansion":  _coerce_str(task_spec.get("role", "an expert assistant")),
        "task":            _coerce_str(task_spec.get("task", task_spec.get("input", ""))),
        "context":         _coerce_str(task_spec.get("context", "")),
        "constraints":     _coerce_str(task_spec.get("constraints", "")),
        "output_format":   _coerce_str(task_spec.get("output_format", "Respond clearly and concisely.")),
        "guidance_blocks": _coerce_str(task_spec.get("guidance_blocks", "")),
        "input":           _coerce_str(task_spec.get("input", "")),
    }

    # Simple str.format replacement (template uses {key} placeholders)
    try:
        rendered = template.format(**subs)
    except KeyError:
        # ponytail: if template has unknown placeholder, render with safe_map — upgrade: use a proper templating lib
        import re
        rendered = re.sub(r"\{(\w+)\}", lambda m: subs.get(m.group(1), ""), template)

    system, user = _split_system_user(rendered)

    # Build params with variant-specific effort overrides
    params = FAMILY_PARAMS[family].copy()
    if model in VARIANT_EFFORT:
        effort_val = VARIANT_EFFORT[model]
        if family == "claude":
            params["effort"] = effort_val
        elif family == "gpt":
            params["reasoning_effort"] = effort_val

    return system, user, params

File: internal/fusion/transforms.go
package fusion

import (
	"fmt"
	"regexp"
	"strings"
)

// familyParams holds per-family sampling defaults.
var familyParams = map[string]map[string]any{
	"claude": {"effort": "high"},
	"gpt":    {"reasoning_effort": "medium"},
	"glm":    {"temperature": 1.0, "top_p": 0.95}, // ponytail: dropped repetition_penalty (llama.cpp-ism, not honored by Z.AI cloud) — values per GLM-5.2 blog (temp=1.0, top_p=0.95)
	"qwen":   {"temperature": 0.7, "top_p": 0.8},
}

// variantEffort maps specific model slugs to effort overrides.
var variantEffort = map[string]string{
	"claude-opus-4-8":   "xhigh",
	"claude-sonnet-4-6": "high",
	"claude-haiku-4-5":  "medium",
	"gpt-5.5":           "high",
	"gpt-5.4":           "medium",
	"gpt-5.4-mini":      "medium",
	"gpt-5.3-codex":     "high",
	"gpt-5.2":           "medium",
	"glm-5.2":           "high", // GLM-5.2 supports High/Max effort via thinking object (verified on coding/paas endpoint)
}

// FamilyOf detects model family from the slug.
// Order matters: qwopus contains "opus" so qwen is checked before claude.
func FamilyOf(model string) string {
	m := strings.ToLower(model)
	switch {
	case strings.Contains(m, "qwen") || strings.Contains(m, "qwopus") ||
		strings.Contains(m, "ollama") || strings.Contains(m, "turboquant"):
		return "qwen"
	case strings.Contains(m, "claude") || strings.Contains(m, "opus") ||
		strings.Contains(m, "sonnet") || strings.Contains(m, "haiku"):
		return "claude"
	case strings.Contains(m, "gpt"):
		return "gpt"
	case strings.Contains(m, "glm"):
		return "glm"
	}
	// ponytail: unknown → claude-style — upgrade: add explicit fallback policy
	return "claude"
}

// TransformResult holds the per-model prompt and sampling params.
type TransformResult struct {
	System string
	User   string
	Params map[string]any
}

// TransformForModel renders a task spec into (system, user, params) for the model's family.
func TransformForModel(spec map[string]any, model string) TransformResult {
	family := FamilyOf(model)
	tmpl, ok := familyTemplates[family]
	if !ok {
		tmpl = familyTemplates["claude"]
	}

	subs := map[string]string{
		"role":            coerceStr(spec["role"], "an expert assistant"),
		"role_expansion":  coerceStr(spec["role"], "an expert assistant"),
		"task":            coerceStr(firstOf(spec, "task", "input"), ""),
		"context":         coerceStr(spec["context"], ""),
		"constraints":     coerceStr(spec["constraints"], ""),
		"output_format":   coerceStr(spec["output_format"], "Respond clearly and concisely."),
		"guidance_blocks": coerceStr(spec["guidance_blocks"], ""),
		"input":           coerceStr(spec["input"], ""),
	}

	rendered := renderTemplate(tmpl, subs)
	system, user := splitSystemUser(rendered)

	params := copyParams(family)
	if effort, ok := variantEffort[model]; ok {
		switch family {
		case "claude":
			params["effort"] = effort
		case "gpt":
			params["reasoning_effort"] = effort
		case "glm":
			// GLM-5.2 effort via thinking object — verified on Z.AI coding/paas endpoint.
			params["thinking"] = map[string]any{"type": "enabled", "effort": effort}
		}
	}

	return TransformResult{System: system, User: user, Params: params}
}

func coerceStr(value any, fallback string) string {
	if value == nil {
		return fallback
	}
	switch v := value.(type) {
	case string:
		if v == "" {
			return fallback
		}
		return v
	case []any:
		parts := make([]string, 0, len(v))
		for _, item := range v {
			parts = append(parts, fmt.Sprintf("- %v", item))
		}
		return strings.Join(parts, "\n")
	default:
		return fmt.Sprintf("%v", v)
	}
}

func firstOf(m map[string]any, keys ...string) any {
	for _, k := range keys {
		if v, ok := m[k]; ok && v != nil {
			return v
		}
	}
	return nil
}

var placeholderRe = regexp.MustCompile(`\{(\w+)\}`)

func renderTemplate(tmpl string, subs map[string]string) string {
	return placeholderRe.ReplaceAllStringFunc(tmpl, func(match string) string {
		key := match[1 : len(match)-1]
		if v, ok := subs[key]; ok {
			return v
		}
		return ""
	})
}

func splitSystemUser(rendered string) (string, string) {
	if idx := strings.Index(rendered, "[USER]"); idx >= 0 {
		system := strings.Replace(rendered[:idx], "[SYSTEM]", "", 1)
		user := strings.TrimSpace(rendered[idx+len("[USER]"):])
		return strings.TrimSpace(system), user
	}
	return strings.TrimSpace(strings.Replace(rendered, "[SYSTEM]", "", 1)), ""
}

func copyParams(family string) map[string]any {
	base, ok := familyParams[family]
	if !ok {
		base = familyParams["claude"]
	}
	out := make(map[string]any, len(base))
	for k, v := range base {
		out[k] = v
	}
	return out
}
